Add a new display_list_benchmarks test suite

gw280 opened the pull request:
This is an initial work in progress for adding a new microbenchmarks suite that'll allow us to get an idea of the relative cost of each raster op defined in our DisplayList format.

gw280 commented:
Sample output from a benchmarking run (currently only `DrawLine` is benchmarked):

```wrightgeorge-macbookpro2:src wrightgeorge$ ./out/host_debug/flow_benchmarks
Run on (16 X 2300 MHz CPU s)
2021-11-05 13:43:01
Benchmark                                             Time           CPU Iterations
-----------------------------------------------------------------------------------
DrawLine<SoftwareCanvasProvider>/16                   6 ms          6 ms        117
DrawLine<SoftwareCanvasProvider>/32                   7 ms          7 ms         96
DrawLine<SoftwareCanvasProvider>/64                  10 ms         10 ms         66
DrawLine<SoftwareCanvasProvider>/128                 18 ms         18 ms         42
DrawLine<SoftwareCanvasProvider>/256                 33 ms         33 ms         21
DrawLine<SoftwareCanvasProvider>/512                 62 ms         62 ms         11
DrawLine<SoftwareCanvasProvider>/1024               120 ms        119 ms          6
DrawLine<SoftwareCanvasProvider>/2k                 262 ms        260 ms          3
DrawLine<MetalCanvasProvider>/16/real_time            8 ms          8 ms         77
DrawLine<MetalCanvasProvider>/32/real_time            8 ms          8 ms         82
DrawLine<MetalCanvasProvider>/64/real_time            8 ms          8 ms         83
DrawLine<MetalCanvasProvider>/128/real_time           8 ms          8 ms         85
DrawLine<MetalCanvasProvider>/256/real_time           7 ms          7 ms         91
DrawLine<MetalCanvasProvider>/512/real_time           8 ms          8 ms         79
DrawLine<MetalCanvasProvider>/1024/real_time          8 ms          8 ms         94
DrawLine<MetalCanvasProvider>/2k/real_time            8 ms          8 ms         94```

gw280 commented:
The basic idea here is to create a unit test for each raster op that:

- Accepts some arguments via the `benchmark::State` object that allows us to modify parameters for the particular raster op we're profiling
- Builds the `DisplayList` to be benchmarked
- Grabs a `SkCanvas` to render that `DisplayList` to using the `CanvasProvider` interface
- Times how long it takes to render that `DisplayList` to the provided `SkCanvas`

Using a templated `CanvasProvider` allows us to separate the logic creating the destination `SkSurface` from the test itself, so we can re-use tests across a Metal, Software, GL, Vulkan `CanvasProvider`. This will give us a good way of benchmarking different backends against each other.

I haven't yet figured out a good way to handle setting attributes for each test, but I am actively working on that. For example, we'd like to be able to run the `DrawLine` tests with both AA enabled and disabled. Jim's work in https://github.com/flutter/engine/pull/29470 should make that easier with the refactoring to use `DisplayListFlags`.

flar commented:
I'm guessing this is all run on the host for now. Do we have a way to run it on some mobile devices as well?

gw280 commented:
> I'm guessing this is all run on the host for now. Do we have a way to run it on some mobile devices as well?

So we don't currently run any of the engine unittests on device. In theory it shouldn't be too hard, but the infrastructure just isn't there right now.

dnfield commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,120 @@
+  // We only want to time the actual rasterization.
+  while (state.KeepRunning()) {
+    displaylist->RenderTo(canvas);
+    canvas->flush();
```
In here you'll wnat to fish out the GrDirectContext and call `flushAndSubmit` or `flush`/`submit` separately.

gw280 commented:
I think this is ready for a first pass at reviewing.

flar reviewed (commented):
I seem to have lost a bunch of comments here during a browser reload, so I'm submitting them as comments (to hopefully bring them back into view) and then resuming the review.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1144 @@
+
+// clang-format off
+
+BENCHMARK_CAPTURE(BM_DrawLine,
```
I did think about that. I've had some thoughts but nothing particularly good. A macro might be the least awful solution.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1144 @@
+    builder.drawLine(SkPoint::Make(i % length, 0),
+                     SkPoint::Make(length - i % length, length));
+    previous = SkPoint::Make(length - i % length, length);
+  }
```
I think this was code that got modified by accident when I was testing some stuff. Deleted all the lines with 'previous' in them.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1144 @@
+  // Always draw 5000 rects regardless of the rect size requested
+  for (size_t i = 0; i < 5000; i++) {
+    // As rects have SkScalar dimensions, we want to ensure that we also
+    // draw non-axis-aligned rects with scalar position and size
```
oh right, I used the phrase axis-aligned when I should have said non-integer coordinates. I'll fix that. The idea here is to have the lines drawn between pixel boundaries to ensure we also benchmark AA.

gw280 replied:
(non-axis-aligned stuff will be tested with the forthcoming attributes tests, in particular setting a rotation matrix)

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1144 @@
+// The resulting image will be an hourglass shape.
+void BM_DrawLine(benchmark::State& state,
+                 std::unique_ptr<CanvasProvider> canvas_provider) {
+  DisplayListBuilder builder;
```
Anything that requires changing the default attributes on the SkPaint will come in a followup PR. For now I just want to focus on the defaults, even if that means the defaults aren't likely to be used that frequently.

gw280 commented on flow/display_list_benchmarks_gl.cc:
```diff
@@ -0,0 +1,189 @@
+
+  sk_sp<SkSurface> MakeOffscreenSurface(const size_t width,
+                                        const size_t height) override {
+    // TODO(gw280): implement this
```
Fixed

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1144 @@
+
+  // Always draw 2000 ovals regardless of the oval size requested. The first
+  // drawOval call draws them shrinking in height, and the second drawOval
+  // call draws them shrinking in width.
```
I just did the one oval around a larger field - I don't think it matters if it's a 3:2 or 2:3 aspect ratio

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1144 @@
+  canvas_provider->Snapshot(filename);
+}
+
+// Draws a series of shrinking square rects from the requested width down
```
Instead of bouncing around I just draw them across the canvas diagonally, but either way it should be fine.

flar reviewed (commented):
About the only functional problem is that there may be overflow on some of the drawImage tests depending on the draw counts compared to the bitmap sizes.

I'm also a little wary of the way that the vertices are constructed leading to degenerate triangles.

Other than that, the rest is just a bunch of nits.

flar commented on flow/display_list_benchmarks_metal.cc:
```diff
@@ -0,0 +1,50 @@
+  sk_sp<SkSurface> MakeOffscreenSurface(const size_t width,
+                                        const size_t height) override {
+    // TODO(wrightgeorge): should we add an API to TestMetalSurface to create an
+    // offscreen?
```
Or maybe it isn't obsolete? Is this a suggestion for a better way to implement this method?

gw280 replied:
Yeah, the suggestion is whether to modify the `TestMetalSurface` class in `flutter/testing`. "Offscreen" is a little bit of a misnomer here anyway because they're all offscreens, but I'm trying to distinguish between the "main" surface and other surfaces. I can delete the comment, I don't think it really matters that much.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+    // draw rects with non-integer position and size
+    builder.drawRect(rect);
+    rect.inset(inset, inset);
+    if (rect.width() <= 5.0f) {
```
stale code, don't remember the exact reasoning.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+// Each vertex colour will alternate through Red, Green, Blue and Cyan.
+sk_sp<SkVertices> GetTestVertices(SkPoint center,
+                                  float radius,
+                                  size_t vertex_count,
```
I think it's better to look at the vertex count because that's the data we actually input. We will need to figure out a way to either grab it, or approximate it. It looks like `SkVertices::approximateSize()` is a half-decent first-order approximation, as it returns an approximation for the total number of bytes used by the `SkVertices` object and all the data arrays. If we just subtract the size of the `SkVertices` object from that then we will have an approximation of the vertex count.

gw280 replied:
We can also grab the vertex count from `Vertices.cc` in `dart:ui`: https://github.com/flutter/engine/blob/main/lib/ui/painting/vertices.cc#L61

It's a little annoying to have to plumb through that count ourselves but I think it's better than trying to calculate the number of triangles.

flar replied:
We'll eventually get to that and flatten all of the Sk structures into the DL, but for now we rely on the data we can grab from the Sk objects.

gw280 replied:
We can't grab the triangle count either, so I think it makes more sense to vary the vertex count to get an idea of that.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+                          : offscreen->makeImageSnapshot();
+    builder.drawImage(
+        image, SkPoint::Make(std::fmod(0.5 * i, bitmap_size), i / bitmap_size),
+        options, true);
```
Yes :) I will add a bounds check, I think at the time I figured it was very unlikely I'd ever set kImagesToDraw that high.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+  //
+  // We don't want to shrink them both at the same time because that would just
+  // degenerate to drawCircle.
+  for (size_t i = 0; i < kOvalsToDraw; i++) {
```
oh, I forgot to push my latest commits. This code is all redone.

gw280 replied:
I'm not sure what you mean by this? I removed the code where we draw two ovals per loop iteration and now we only draw one. I kept the loop condition at 1000 because I think 2000 is probably too many anyway.

flar replied:
I misread. Sorry.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+
+  // We only want to time the actual rasterization.
+  for (auto _ : state) {
+    canvas->discard();
```
old artifact, I don't think it's necessary. I'll remove them.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+
+  SkRect bounds = SkRect::MakeLTRB(0, 0, length, length);
+
+  for (size_t i = 0; i < kArcsToDraw; i++) {
```
Will fix.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+  for (size_t i = 1; i < sides; i++) {
+    path.lineTo(points[i]);
+  }
+  path.lineTo(points[0]);
```
Yes, I had it there for symmetry for that exact reason. I'd prefer an explicit lineTo. It also means the path verb counts are consistent (if we used close => lineTo(), then LinesPath would have 1 fewer verbs per path ).

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+}
+
+// Generate `count` pseudo-random test points that are arbitrarily chosen.
+// The points generated will wrap in x and y for the bounds of `canvas_size`.
```
Fixed.

gw280 commented on flow/display_list_benchmarks_gl.cc:
```diff
@@ -0,0 +1,56 @@
+    const auto image_info = SkImageInfo::MakeN32Premul(surface_size_);
+    surface_ = SkSurface::MakeRenderTarget(
+        gl_surface_->GetGrContext().get(), SkBudgeted::kNo, image_info, 1,
+        kTopLeft_GrSurfaceOrigin, nullptr, false);
```
I'm not sure, but it won't hurt to add.

gw280 commented on flow/display_list_benchmarks.cc:
```diff
@@ -0,0 +1,1000 @@
+        vertices.push_back(outer_points[i % outer_points.size()]);
+        colors.push_back(i % 2 ? SK_ColorRED : SK_ColorGREEN);
+        vertices.push_back(center);
+        colors.push_back(i % 2 ? SK_ColorBLUE : SK_ColorCYAN);
```
Oh yes, this is because the vertices were originally set up to do a doughnut shape and I just replaced the inner coordinates with the center point every time. I'll rework it.

gw280 commented:
I've rebased this branch off the latest `main`, which includes Chinmay's changes to move the DisplayList code out of `flow`. As a result I've squashed the history and renamed the benchmark suite to `display_list_benchmarks`.

The only review comment that still needs to be addressed is the `DrawVertices` stuff.

flar reviewed (approved):
LGTM - some comments about work that you can consider or not or do at a later date, but nothing blocking.
